initial implementation of footnotes extension

3b opened the pull request:
inline references like `[^foo]`, with definitions as blocks starting with `[^foo]:`.

questions for users::
* bugs/missing features?
* what should be configurable and/or have different defaults?
    * css classes?
    * formatting of links?
    * container element for list of footnotes?
    * separator before/after definitions?
* is `^[inline footnote]` version common/useful enough to be worth implementing? (or some other syntax)


todo:
* [x] text/md output
* [ ] documentation
* [ ] configuration
* [x] tests
* [ ] improve formatting of output?
* [ ] option for adding some extra text to footnote IDs to allow displaying multiple markdown documents on one page (for example comments, or making an html file out of the test suite)

3b committed:
initial implementation of footnotes extension

inline references like `[^foo]`, with definitions as blocks
starting with `[^foo]:`.

3b committed:
rename file to match .asd

scymtym commented on 3bmd-footnote.lisp:
```diff
@@ -0,0 +1,187 @@
+
+;;; we parse footnote definitions basically like a loose list, but
+;;; without a "next list item"
+(defrule footnote-block (and (! 3bmd-grammar::blank-line)
+                             3bmd-grammar::line
+                             (* footnote-block-line))
+  (:destructure (b l block)
+    (declare (ignore b))
+    (text l block)))
```
The rule action can be written as
```cl
(:function rest)
(:destructure (l block)
  (text l block))
```

scymtym commented on 3bmd-footnote.lisp:
```diff
@@ -0,0 +1,187 @@
+                                  3bmd-grammar::optionally-indented-line)
+  (:destructure (i1 i2 line)
+    (declare (ignore i1 i2))
+    line))
```
This is just `(:function third)`.

scymtym commented on 3bmd-footnote.lisp:
```diff
@@ -0,0 +1,187 @@
+    (and footnote (! #\:))
+  (:destructure (id x)
+                (declare (ignore x))
+                (list 'footnote-ref id)))
```
For rule actions, esrap can compose multiple transformations. In cases like this, it is convenient to "pre-process" with a `:function` action:
```cl
(:function first)
(:lambda (id)
  (list 'footnote-ref id))
```

scymtym commented on 3bmd-footnote.lisp:
```diff
@@ -0,0 +1,187 @@
+                       "]")
+  (:destructure (o id c)
+    (declare (ignore o c))
+    (text id)))
```
See next comment for general explanation.
```cl
(:function second)
(:text t)
```

scymtym reviewed (commented):
Mentioned a few simplifications using composition of esrap rule actions.

3b committed:
simplifications for parsing rules from @scymtym

3b committed:
add :markdown and :plain output formats for footnotes

Currently prints definitions where they were in original for both, not sure
if it would be better to move to the end like html?

3b committed:
remove debug print

3b committed:
start adding tests/fixing bugs

3b committed:
rename footnotes package to better match system/flag and other extensions

3b committed:
add some documentation for footnotes extension
